### PR TITLE
Revised inner product

### PR DESCRIPTION
**Pull request by harsha-simhadri**

### CODE REVIEW

File: include/utils.h [hunk 1]
```diff
@@ -410,6 +418,71 @@ namespace diskann {
     }
   }
```
Comment (harsha-simhadri) [inline include/utils.h:420]: Comment on the vector transformation

File: src/aux_utils.cpp [hunk 2]
```diff
@@ -354,20 +354,20 @@ namespace diskann {
     double full_index_ram =
         ESTIMATE_RAM_USAGE(base_num, base_dim, sizeof(T), R);
     if (full_index_ram < ram_budget * 1024 * 1024 * 1024) {
-      diskann::cout << "Full index fits in RAM, building in one shot"
+      diskann::cout << "Full index fits in RAM budget, should consume at most " << full_index_ram/(1024*1024*1024) <<"GBs, so building in one shot"
```
Comment (harsha-simhadri) [inline src/aux_utils.cpp:357]: GB or GiB?

File: src/aux_utils.cpp [hunk 3]
```diff
@@ -612,25 +621,46 @@ namespace diskann {
...
-             "indexing)"
+             "indexing) B' (PQ bytes for disk index: optional parameter for very large dimensional data)"
           << std::endl;
       return false;
     }
+  
+
+    if (!std::is_same<T, float>::value && compareMetric == diskann::Metric::INNER_PRODUCT) {
+              std::stringstream stream;
+        stream << "DiskANN currently only supports floating point data for Max Inner Product Search. Please contact us if you need other scenarios." << std::endl;
```
Comment (harsha-simhadri) [inline src/aux_utils.cpp:646]: Drop contacT?

File: src/aux_utils.cpp [hunk 4]
```diff
@@ -640,6 +670,17 @@ namespace diskann {
     std::string medoids_path = disk_index_path + "_medoids.bin";
     std::string centroids_path = disk_index_path + "_centroids.bin";
     std::string sample_base_prefix = index_prefix_path + "_sample";
+    std::string disk_pq_pivots_path = index_prefix_path + "_disk.index_pq_pivots.bin"; // optional if disk index is also storing pq data
+    std::string disk_pq_compressed_vectors_path = // optional if disk index is also storing pq data
+        index_prefix_path + "_disk.index_pq_compressed.bin";
+
+// output a new base file which contains extra dimension with sqrt(1 - ||x||^2/M^2) for every x, M is max norm of all points. Extra space on disk needed!
+    if (compareMetric == diskann::Metric::INNER_PRODUCT) {
+      std::cout<<"Using Inner Product search, so need to pre-process base data into temp file. Please ensure there is sufficient space!!" << std::endl;
```
Comment (harsha-simhadri) [inline src/aux_utils.cpp:679]: CAn we be specific about extra space? Say create another file as large as base set?

File: src/index.cpp [hunk 5]
```diff
@@ -501,9 +504,9 @@ namespace diskann {
       std::vector<unsigned>     init_ids,
       std::vector<Neighbor> &   expanded_nodes_info,
       tsl::robin_set<unsigned> &expanded_nodes_ids) {
-    const T *             node_coords = _data + _aligned_dim * node_id;
+    T *             node_coords = _data + _aligned_dim * node_id;
```
Comment (harsha-simhadri) [inline src/index.cpp:507]: add const?

File: src/index.cpp [hunk 6]
```diff
@@ -535,7 +538,7 @@ namespace diskann {
     float cur_alpha = 1;
     while (cur_alpha <= alpha && result.size() < degree) {
       unsigned start = 0;
-
+      float eps = cur_alpha + 0.01;
```
Comment (harsha-simhadri) [inline src/index.cpp:541]: comment on the reasoning here?

File: src/partition_and_pq.cpp [hunk 7]
```diff
@@ -221,24 +225,27 @@ int generate_pq_pivots(const float *passed_train_data, size_t num_train,
...
+  }
+    if (make_zero_mean) {
+    for (uint64_t d = 0; d < dim; d++) {
     for (uint64_t p = 0; p < num_train; p++) {
       centroid[d] += train_data[p * dim + d];
     }
     centroid[d] /= num_train;
   }
 
   //  std::memset(centroid, 0 , dim*sizeof(float));
```
Comment (harsha-simhadri) [inline src/partition_and_pq.cpp:242]: do we need memset?

File: src/partition_and_pq.cpp [hunk 8]
```diff
@@ -221,24 +225,27 @@ int generate_pq_pivots(const float *passed_train_data, size_t num_train,
...
     }
   }
-
+  
   // Calculate centroid and center the training data
   std::unique_ptr<float[]> centroid = std::make_unique<float[]>(dim);
   for (uint64_t d = 0; d < dim; d++) {
     centroid[d] = 0;
+  }
+    if (make_zero_mean) {
```
Comment (harsha-simhadri) [inline src/partition_and_pq.cpp:234]: add a comment on when we might need zero mean

File: src/partition_and_pq.cpp [hunk 9]
```diff
@@ -751,9 +937,9 @@ int partition(const std::string data_file, const float sampling_rate,
   // now pivots are ready. need to stream base points and assign them to
   // closest clusters.
 
-  std::vector<size_t> cluster_sizes;
-  estimate_cluster_sizes<T>(data_file, pivot_data, num_parts, train_dim, k_base,
-                            cluster_sizes);
+  //std::vector<size_t> cluster_sizes;
+  //estimate_cluster_sizes<T>(data_file, pivot_data, num_parts, train_dim, k_base,
+  //                          cluster_sizes);
```
Comment (harsha-simhadri) [inline src/partition_and_pq.cpp:942]: remove?

File: tests/build_disk_index.cpp [hunk 10]
```diff
@@ -11,29 +11,32 @@
...
-      dataFilePath, indexFilePath, indexBuildParameters, diskann::Metric::L2);
+      dataFilePath, indexFilePath, indexBuildParameters, metric);
 }
 
 int main(int argc, char** argv) {
-  if (argc != 9) {
+  if (argc != 11) {
     std::cout << "Usage: " << argv[0]
-              << "  [data_type<float/int8/uint8>]  [data_file.bin]  "
+              << "  [data_type<float/int8/uint8>]  [dist_fn: 0 for L2, 1 for MIPS] [data_file.bin]  "
```
Comment (harsha-simhadri) [inline tests/build_disk_index.cpp:22]: Why not ask for "L2" or "MIPS"

File: tests/search_disk_index.cpp [hunk 11]
```diff
@@ -196,7 +208,7 @@ int search_disk_index(int argc, char** argv) {
 
   uint32_t optimized_beamwidth = 2;
 
-  //  query_num = 1;
+//query_num = 1;
```
Comment (harsha-simhadri) [inline tests/search_disk_index.cpp:211]: remove?

File: tests/search_memory_index.cpp [hunk 12]
```diff
@@ -160,11 +169,10 @@ int main(int argc, char** argv) {
   if (argc < 11) {
     std::cout
         << "Usage: " << argv[0]
-        << "  [index_type<float/int8/uint8>]  [data_file.bin]  "
+        << "  [index_type<float/int8/uint8>]  [dist_fn (0 for L2, 1 for Inner Product, 2 for Fast L2 for small datasets)] [data_file.bin]  "
```
Comment (harsha-simhadri) [inline tests/search_memory_index.cpp:172]: update help string to reflect "L2/MIPS"

File: tests/utils/compute_groundtruth.cpp [hunk 13]
```diff
@@ -104,6 +104,32 @@ void distsq_to_points(
...
+  bool ones_vec_alloc = false;
+  if (ones_vec == NULL) {
+    ones_vec = new float[nqueries > npoints ? nqueries : npoints];
+    std::fill_n(ones_vec, nqueries > npoints ? nqueries : npoints, (float) 1.0);
+    ones_vec_alloc = true;
+  }
+  cblas_sgemm(CblasColMajor, CblasTrans, CblasNoTrans, npoints, nqueries, dim,
+              (float) -1.0, points, dim, queries, dim, (float) 0.0, dist_matrix,
+              npoints);
+  //  cblas_sgemm(CblasColMajor, CblasNoTrans, CblasTrans, npoints, nqueries, 1,
```
Comment (harsha-simhadri) [inline tests/utils/compute_groundtruth.cpp:123]: remove comment

File: tests/utils/compute_groundtruth.cpp [hunk 14]
```diff
@@ -303,6 +344,23 @@ int aux_main(int argv, char **argc) {
 
     delete[] closest_points_part;
     delete[] dist_closest_points_part;
+
+    /*
```
Comment (harsha-simhadri) [inline tests/utils/compute_groundtruth.cpp:348]: remove

File: include/aux_utils.h [hunk 15]
```diff
@@ -29,13 +29,16 @@ typedef int FileHandle;
 #include "common_includes.h"
 #include "utils.h"
 #include "windows_customizations.h"
+#include "gperftools/malloc_extension.h"
+
 
 namespace diskann {
-  const size_t   TRAINING_SET_SIZE = 1500000;
```
[the remark this answers is not in the document]
Comment (rakri) [reply include/aux_utils.h:34]: Let's make it larger, to a million. If necessary we can make it small.

File: include/distance.h [hunk 16]
```diff
@@ -302,9 +307,14 @@ namespace diskann {
     virtual float compare(const float *a, const float *b,
                           unsigned int length) const {
 #ifndef _WINDOWS
```
[the remark this answers is not in the document]
Comment (rakri) [reply include/distance.h:309]: it should be ifdef LINUX, but we need a much bigger change, so leaving as is.

File: include/distance.h [hunk 17]
```diff
@@ -328,7 +338,7 @@ namespace diskann {
   template<typename T>
   class DistanceInnerProduct : public Distance<T> {
    public:
-    float compare(const T *a, const T *b, unsigned size) const {
+    float acompare(const T *a, const T *b, unsigned size) const {
```
[the remark this answers is not in the document]
Comment (rakri) [reply include/distance.h:341]: will do.

File: include/partition_and_pq.h [hunk 18]
```diff
@@ -54,7 +62,7 @@ DISKANN_DLLEXPORT int generate_pq_pivots(const float *train_data,
                                          unsigned    num_centers,
                                          unsigned    num_pq_chunks,
                                          unsigned    max_k_means_reps,
-                                         std::string pq_pivots_path);
+                                         std::string pq_pivots_path, bool make_zero_mean = false);
```
[the remark this answers is not in the document]
Comment (rakri) [reply include/partition_and_pq.h:65]: ok, will compile with clang, hopefully that does it.

File: include/pq_table.h [hunk 19]
```diff
@@ -150,5 +148,61 @@ namespace diskann {
       }
     }
   }
+
+ float compare(const float* query_vec, _u8* base_vec) {
```
[the remark this answers is not in the document]
Comment (rakri) [reply include/pq_table.h:152]: done

File: include/utils.h (hunk 1 above)
Comment (rakri) [reply include/utils.h:420]: done.

File: src/aux_utils.cpp (hunk 2 above)
Comment (rakri) [reply src/aux_utils.cpp:357]: Fixed. GiB.

File: src/aux_utils.cpp (hunk 3 above)
Comment (rakri) [reply src/aux_utils.cpp:646]: Ok.

File: src/aux_utils.cpp (hunk 4 above)
Comment (rakri) [reply src/aux_utils.cpp:679]: done. we need a note somewhere on total interim consumption also for all mem indices.

File: src/index.cpp (hunk 5 above)
Comment (rakri) [reply src/index.cpp:507]: ok. done.

File: src/index.cpp (hunk 6 above)
Comment (rakri) [reply src/index.cpp:541]: added.

File: src/partition_and_pq.cpp (hunk 7 above)
Comment (rakri) [reply src/partition_and_pq.cpp:242]: removed.

File: src/partition_and_pq.cpp (hunk 8 above)
Comment (rakri) [reply src/partition_and_pq.cpp:234]: done

File: src/partition_and_pq.cpp (hunk 9 above)
Comment (rakri) [reply src/partition_and_pq.cpp:942]: done

File: tests/utils/compute_groundtruth.cpp (hunk 13 above)
Comment (rakri) [reply tests/utils/compute_groundtruth.cpp:123]: done.

File: tests/utils/compute_groundtruth.cpp (hunk 14 above)
Comment (rakri) [reply tests/utils/compute_groundtruth.cpp:348]: done.

File: tests/build_disk_index.cpp (hunk 10 above)
Comment (rakri) [reply tests/build_disk_index.cpp:22]: done.

File: tests/search_disk_index.cpp (hunk 11 above)
Comment (rakri) [reply tests/search_disk_index.cpp:211]: done.

File: tests/search_memory_index.cpp (hunk 12 above)
Comment (rakri) [reply tests/search_memory_index.cpp:172]: done.

File: tests/utils/vector_analysis.cpp [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+int analyze_norm(std::string base_file) {
+  std::cout << "Analyzing data norms" << std::endl;
+  T*   data;
+  _u64 npts, ndims;
+  diskann::load_bin<T>(base_file, data, npts, ndims);
+  std::vector<float> norms(npts, 0);
+#pragma omp parallel for schedule(dynamic)
+  for (_u32 i = 0; i < npts; i++) {
+    for (_u32 d = 0; d < ndims; d++)
+      norms[i] += data[i * ndims + d] * data[i * ndims + d];
```
Comment (ShikharJ) [inline tests/utils/vector_analysis.cpp:33]: Doesn't this need a sqrt()?

File: tests/utils/vector_analysis.cpp [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+  return 0;
+}
+
+int main(int argc, char** argv) {
+  if (argc < 4) {
+    std::cout
+        << argv[0]
+        << " data_type [float/int8/uint8] base_bin_file "
+           "[option: 1-norm analysis, 2-prep_base_for_mip, "
+           "3-prep_query_for_mip, 4-normalize-vecs] [out_file for options 2/3]"
```
Comment (ShikharJ) [inline tests/utils/vector_analysis.cpp:133]: This should be `out_file for options 2/3/4` I think.

Comment (harsha-simhadri) [reply tests/utils/vector_analysis.cpp:133]: done

File: tests/search_memory_index.cpp [hunk 22]
```diff
@@ -160,11 +175,11 @@ int main(int argc, char** argv) {
   if (argc < 11) {
     std::cout
         << "Usage: " << argv[0]
-        << "  [index_type<float/int8/uint8>]  [data_file.bin]  "
+        << "  [index_type<float/int8/uint8>]  [dist_fn (l2/mips/fast_l2)] "
+           "[data_file.bin]  "
            "[memory_index_path]  [num_threads] "
            "[query_file.bin]  [truthset.bin (use \"null\" for none)] "
-           " [K]  [result_output_prefix] [use_optimized_search (for small ~1M "
-           "data)] "
```
Comment (harsha-simhadri) [inline tests/search_memory_index.cpp:167]: @rakri, are we dropping support for "1M size optimized index"

**Pull request merged**